### PR TITLE
Fix inspection of unspecified args for container hparams

### LINKED ISSUE #8948: save_hyperparameters attempts to parse unspecified args when a namespace is specified
## 🐛 Bug
save_hyperparameters attempts to parse unspecified args when a namespace is specified

### To Reproduce

```python
import argparse

import pandas as pd
import pytorch_lightning as pl

class plmodule(pl.LightningModule):
    def __init__(self, hp, df):
        super().__init__()
        self.save_hyperparameters(hp)


hp = argparse.Namespace(**{"hello": "world"})
df = pd.DataFrame(["hello", "world"])
x = plmodule(hp, df)
```

### Trace
<details>
  <summary>Expand</summary>

```python
ValueError                                Traceback (most recent call last)
<ipython-input-2-8ea9fd40a08a> in <module>
     12 hp = argparse.Namespace(**{"hello": "world"})
     13 df = pd.DataFrame(["hello", "world"])
---> 14 x = plmodule(hp, df)

<ipython-input-2-8ea9fd40a08a> in __init__(self, hp, df)
      7     def __init__(self, hp, df):
      8         super().__init__()
----> 9         self.save_hyperparameters(hp)
     10 
     11 

/opt/conda/lib/python3.8/site-packages/pytorch_lightning/core/mixins/hparams_mixin.py in save_hyperparameters(self, ignore, frame, logger, *args)
    103         if not frame:
    104             frame = inspect.currentframe().f_back
--> 105         save_hyperparameters(self, *args, ignore=ignore, frame=frame)
    106 
    107     def _set_hparams(self, hp: Union[dict, Namespace, str]) -> None:

/opt/conda/lib/python3.8/site-packages/pytorch_lightning/utilities/parsing.py in save_hyperparameters(obj, ignore, frame, *args)
    240         if len(isx_non_str) == 1:
    241             hp = args[isx_non_str[0]]
--> 242             cand_names = [k for k, v in init_args.items() if v == hp]
    243             obj._hparams_name = cand_names[0] if cand_names else None
    244         else:

/opt/conda/lib/python3.8/site-packages/pytorch_lightning/utilities/parsing.py in <listcomp>(.0)
    240         if len(isx_non_str) == 1:
    241             hp = args[isx_non_str[0]]
--> 242             cand_names = [k for k, v in init_args.items() if v == hp]
    243             obj._hparams_name = cand_names[0] if cand_names else None
    244         else:

/opt/conda/lib/python3.8/site-packages/pandas/core/generic.py in __nonzero__(self)
   1532     @final
   1533     def __nonzero__(self):
-> 1534         raise ValueError(
   1535             f"The truth value of a {type(self).__name__} is ambiguous. "
   1536             "Use a.empty, a.bool(), a.item(), a.any() or a.all()."

ValueError: The truth value of a DataFrame is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().
```
  
</details>

### Expected behavior
module is created with hparams `"hello": world`

### Environment

- PyTorch Lightning Version: reproduced in 1.3.7, 1.3.8, 1.4.2
- PyTorch Version: 1.8.0a0+52ea372
- Python version: 3.8

### Additional context
Seems to be an old bug and not a new regression, setting `df = 1` results in the correct behavior, but `save_hyperparameters` shouldnt be accessing other args when a namespace is already specified.

Encountered this as I was trying to set hparams in the datamodule as changed in 1.4 #3792

Relevant code block:
https://github.com/PyTorchLightning/pytorch-lightning/blob/938a191406fff5f51fba03fcf824f22d8d23c2e0/pytorch_lightning/utilities/parsing.py#L203

### PR DESCRIPTION
**Pull request by s-rog**
## What does this PR do?

Fixes #8948

### Does your PR introduce any breaking changes? If yes, please list them.

The internal logic for qualifying containers has been changed from `not isinstance(str)` to specific types (Namespace, dict, omegaconf if available). May break undocumented edge case uses.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

## Todo:

- [x] omegaconf
- [x] tests

the tests only test for functionality, not sure how I would test for "non-inspection" besides passing in a type that breaks the parsing which is not ideal.

### CODE REVIEW

File: pytorch_lightning/utilities/parsing.py [hunk 1]
```diff
@@ -204,46 +204,57 @@ def save_hyperparameters(
...
-            ignore = [arg for arg in ignore if isinstance(arg, str)]
-        init_args = {k: v for k, v in init_args.items() if k not in ignore}
-
-    if not args:
-        # take all arguments
-        hp = init_args
-        obj._hparams_name = "kwargs" if hp else None
+    # container
+    elif len(args) == 1 and isinstance(args[0], (Namespace, dict)):
+        # todo: omegaconf?
```
Comment (tchaton) [inline pytorch_lightning/utilities/parsing.py:212]: Mind adressing this comment in this PR ?

File: pytorch_lightning/utilities/parsing.py [hunk 2]
```diff
@@ -204,46 +204,57 @@ def save_hyperparameters(
...
+        hp = args[0]
+        if isinstance(args[0], dict):
+            hp = Namespace(**args[0])
+        obj._hparams_name = "hparams"
+        obj._set_hparams(hp)
+        obj._hparams_initial = copy.deepcopy(obj._hparams)
+        return
+    # everything else
```
Comment (tchaton) [inline pytorch_lightning/utilities/parsing.py:220]: Mind putting more. detailed comments.

**Review by tchaton — COMMENTED** [review]
Hey @s-rog. Mind putting more comments on the code. This part of the code might be complex for new contributors or reviewers.

File: pytorch_lightning/utilities/parsing.py (hunk 2 above)
Comment (s-rog) [inline pytorch_lightning/utilities/parsing.py:220]:
```suggestion
    # non-container args parsing
```

File: pytorch_lightning/utilities/parsing.py [hunk 3]
```diff
@@ -204,46 +204,57 @@ def save_hyperparameters(
...
-    if not args:
-        # take all arguments
-        hp = init_args
-        obj._hparams_name = "kwargs" if hp else None
+    # container
+    elif len(args) == 1 and isinstance(args[0], (Namespace, dict)):
+        # todo: omegaconf?
+        hp = args[0]
+        if isinstance(args[0], dict):
+            hp = Namespace(**args[0])
```
Comment (awaelchli) [inline pytorch_lightning/utilities/parsing.py:215]: why do convert to Namespace here?

Comment (s-rog) [reply pytorch_lightning/utilities/parsing.py:215]: It is redundant and shall be removed

File: pytorch_lightning/utilities/parsing.py (hunk 1 above)
Comment (s-rog) [reply pytorch_lightning/utilities/parsing.py:212]: Overlooked this when I was doing the bug fixes sorry about that.

I'm hesitant to import omegaconf here (to add to `isinstance`), while the previous implementation only checks for `len(args) == 1` and `not isinstance(args[0], str)` which is quite ambiguous. Perhaps there is another check I missed? or is it fine to import omegaconf here?

**Review by awaelchli — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/utilities/parsing.py [hunk 4]
```diff
@@ -204,46 +208,57 @@ def save_hyperparameters(
...
+            if current_frame:
+                frame = current_frame.f_back
+        if not isinstance(frame, types.FrameType):
+            raise AttributeError("There is no `frame` available while being required.")
+
+        if is_dataclass(obj):
+            init_args = {f.name: getattr(obj, f.name) for f in fields(obj)}
+        else:
+            init_args = get_init_args(frame)
+        assert init_args, "failed to inspect the obj init"
```
Comment (Borda) [inline pytorch_lightning/utilities/parsing.py:239]:
```suggestion
        assert init_args, f"failed to inspect the obj init - {frame}"
```

File: pytorch_lightning/utilities/parsing.py [hunk 5]
```diff
@@ -204,46 +208,57 @@ def save_hyperparameters(
...
+
+        if ignore is not None:
+            if isinstance(ignore, str):
+                ignore = [ignore]
+            if isinstance(ignore, (list, tuple)):
```
Comment (Borda) [inline pytorch_lightning/utilities/parsing.py:244]:
```suggestion
            if isinstance(ignore, (list, tuple, set)):
```

**Review by Borda — APPROVED** [review]

**Comment by awaelchli** [comment]
@s-rog we may have missed a detail here regarding the _hparams_name, see #9631. Let's see if we can cover this edge case.

File: pytorch_lightning/utilities/parsing.py [hunk 6]
```diff
@@ -204,46 +208,57 @@ def save_hyperparameters(
...
-        init_args = {k: v for k, v in init_args.items() if k not in ignore}
-
-    if not args:
-        # take all arguments
-        hp = init_args
-        obj._hparams_name = "kwargs" if hp else None
+    # container
+    elif len(args) == 1 and isinstance(args[0], tuple(hparams_container_types)):
+        hp = args[0]
+        obj._hparams_name = "hparams"
```
Comment (awaelchli) [inline pytorch_lightning/utilities/parsing.py:220]: here it is immediately assume that the name of the arg was "hparams". previously, the logic below was checking for matching argument names in the signature. 

While this old way of passing in hyperparameters is not our recommended way, we have not officially dropped backward compatibility for this and therefore this was an unwanted breaking change imo.

Comment (awaelchli) [reply pytorch_lightning/utilities/parsing.py:220]: See #9631 for an example of what was working previously.

**Pull request merged**